### PR TITLE
Refactor handling of Picard command-line args

### PR DESCRIPTION
**Pull request by skelly37**
# Summary

* This is a…
  * [ ] Bug fix
  * [ ] Feature addition
  * [x] Refactoring
  * [ ] Minor / simple change (like a typo)
  * [ ] Other
* **Describe this change in 1-2 sentences**:
I have removed the need to duplicate the arguments passed during Tagger's initialization

# Problem

After merging the PR introducing the command-line arguments processing, Tagger now gets both unparsed arguuments & parsed items to load by the app (files, MBIDs, URLs & commands)

* JIRA ticket (_optional_): None

# Solution

- Merge command-line arguments and remote commands into `picard_args`.
- Name pipe with an unique identifier

# Action

### CODE REVIEW

**Comment by skelly37** [comment]
TODO: Support multiple arguments in commands like `picard -e LOAD`
REF: https://github.com/metabrainz/picard/pull/2137#issuecomment-1230697994

**Comment by skelly37** [comment]
@zas @phw Enjoy!

@rdswift I have also included your suggestion here so you might want to take a look

**Comment by skelly37** [comment]
Everything after [d329ff0](https://github.com/metabrainz/picard/pull/2139/commits/d329ff0df5b3a5efb587775df0d5232b5ef54aac) (except [6a70deb](https://github.com/metabrainz/picard/pull/2139/commits/6a70debde6ffd9eb497d540226fde15479ad448c) of course) is a questionable suggestion, I can remove some commit, no hard feelings

File: picard/tagger.py
Comment (zas) [inline picard/tagger.py:1065]: I think this change is incorrect, our log.* functions support passing format arguments but here that's a Qt message box.

File: picard/tagger.py
Comment (zas) [inline picard/tagger.py:1267]: Why capitalized attribute's names?

I have the feeling we could subclass `argparse_args` object and extends it, rather than copying all of its attributes.

File: picard/tagger.py
Comment (zas) [inline picard/tagger.py:1299]: Why `del` those attributes?

File: picard/tagger.py
Comment (zas) [inline picard/tagger.py:1402]: Why `del`?

File: picard/tagger.py
Comment (zas) [inline picard/tagger.py:1429]: Why `del`?

File: picard/tagger.py
Comment (zas) [inline picard/tagger.py:1444]: Why `del`?

**Review by zas — CHANGES_REQUESTED** [review]

File: picard/tagger.py
Comment (skelly37) [reply picard/tagger.py:1065]: Right, my overlook

File: picard/tagger.py
Comment (skelly37) [reply picard/tagger.py:1267]: >Why capitalized attribute's names?
To show they're constant


>I have the feeling we could subclass argparse_args object and extends it, rather than copying all of its attributes.
I think it would compound the readability and simplicity of this class

File: picard/tagger.py
Comment (skelly37) [reply picard/tagger.py:1299]: Not needed further, minor memory optimization, as all other `del`s. What you're trying to say that explicit freeing the memory ASAP is not the pythonic way, right? But they are really not needed during runtime, yet they'd be deleted after the GUI is closed (unless I misunderstand something in Python's memory management). The same goes for all other `del`s

Comment (zas) [reply picard/tagger.py:1299]: Well, `del` doesn't free memory by itself. That's not like a C `free()`. It marks objects as unused, and memory will eventually be freed when Python garbage collector runs.

It makes sense to call `del` to remove an unused variable in some cases, but IMHO you shouldn't bother too much (but perhaps if the variable contains a LOT of data). Or move the variable to a function, call the function, and variable will be freed on return.

File: picard/tagger.py
Comment (zas) [reply picard/tagger.py:1267]: Even though there is no strict definition of a constant in Python (well, they are _usually_ defined at module level), those are normal class attributes, so I don't think they should be capitalized.
A class constant can be capitalized, but it is usually never modified, that's not the case here for `PROCESSABLE` which is initialized then modified.
I don't think upper-cased names add anything here.

File: picard/tagger.py
Comment (rdswift) [inline picard/tagger.py:1284]: By defaulting to an empty list rather than `['']` I think that any commands without arguments such as `-e CLUSTER_ALL` will not be added to `self.PROCESSABLE`.  That's why my suggestion was to default to `['']`.

File: picard/tagger.py
Comment (zas) [inline picard/tagger.py:1251]: Is there a reason to make `argparse_args` optional?

File: picard/tagger.py
Comment (zas) [inline picard/tagger.py:1255]: This method is only called from `__init__()` so the code can be moved there and this method removed.

**Review by zas — CHANGES_REQUESTED** [review]

File: picard/tagger.py
Comment (skelly37) [reply picard/tagger.py:1251]: Well, actually no.

File: po/appstream/en_CA.po
Comment (zas) [inline po/appstream/en_CA.po:12]: This shouldn't have been changed ;)

**Review by zas — CHANGES_REQUESTED** [review]

File: picard/tagger.py
Comment (zas) [inline picard/tagger.py:1254]: Why is this disappeared?

File: picard/tagger.py
Comment (zas) [inline picard/tagger.py:1250]: Why is this disappeared?
Those methods are required.

**Review by zas — CHANGES_REQUESTED** [review]

File: po/appstream/en_CA.po
Comment (skelly37) [reply po/appstream/en_CA.po:12]: A little fckup provided by PyCharm (: Gimme 1min

File: picard/tagger.py
Comment (skelly37) [reply picard/tagger.py:1254]: Where is it used? I cannot find any usage

File: picard/tagger.py
Comment (skelly37) [reply picard/tagger.py:1250]: Where is it used? I cannot find any usage

Comment (skelly37) [reply picard/tagger.py:1250]: So basically that's why I moved the logic to `PicardArgs` :)

Comment (rdswift) [reply picard/tagger.py:1250]: I believe that move is also the reason for the failing tests.

Comment (zas) [reply picard/tagger.py:1250]: > Where is it used? I cannot find any usage

On master:

```bash
❯ python tagger.py -V             
Picard 2.9.0.dev1, Python 3.10.4, PyQt 5.15.7, Qt 5.15.2, Mutagen 1.45.1, Discid discid 1.2.0, libdiscid 0.6.2, astrcmp Python, SSL OpenSSL 1.1.1m  14 Dec 2021
❯ python tagger.py -v
MusicBrainz Picard 2.9.0.dev1

```

On this branch, behavior changed and Picard starts instead.

Comment (rdswift) [reply picard/tagger.py:1250]: The problem is that you changed lines 1353-1355.

Comment (skelly37) [reply picard/tagger.py:1250]: Okay, I see that I forgot to assign the value to the `PicardArgs.version` and it always remained the default `""` empty string

Comment (rdswift) [reply picard/tagger.py:1250]: What output do you get when you start it with the `--long-version ` command line option?

Comment (skelly37) [reply picard/tagger.py:1250]: @rdswift It works exactly the same for the end user, but the functions are no longer stand-alone code but wrapped in the class.

Comment (skelly37) [reply picard/tagger.py:1250]: See PicardArgs. __get_version(), PicardArgs.version value depends on the flag provided by the user.

Comment (rdswift) [reply picard/tagger.py:1250]: Okay, thanks.

Comment (rdswift) [reply picard/tagger.py:1250]: Thinking about this a bit more...  Not that it makes a big difference, but I sort of question the need to store the version string in a variable when it is only used once.  Then again, maybe it is stored as an empty string if neither `-v` nor `-V` are specified on the command line, and Picard exits after displaying the string when `-v` or `-V` are specified.

Comment (skelly37) [reply picard/tagger.py:1250]: > I sort of question the need to store the version string in a variable when it is only used once. Then again, maybe it is stored as an empty string if neither -v nor -V are specified on the command line, and Picard exits after displaying the string when -v or -V are specified.

If we wanted to introduce just `PicardArgs.get_version()`, we'd end up having to store the booleans from `argparse` to determine which wersion actually return, so we still end up with 2 variables. And since @zas advised me not to overuse `del`, I left it there.

File: picard/tagger.py
Comment (zas) [inline picard/tagger.py:1283]:
```suggestion
                args = e[1:] or ['']
```

```python
>>> e = []
>>> e[1:] if len(e) > 1 else ['']
['']
>>> e[1:] or ['']
['']
>>> e = ['a']
>>> e[1:] if len(e) > 1 else ['']
['']
>>> e[1:] or ['']
['']
>>> e = ['a', 'b']
>>> e[1:] if len(e) > 1 else ['']
['b']
>>> e[1:] or ['']
['b']
```

File: picard/tagger.py
Comment (zas) [inline picard/tagger.py:1288]: Those `del` are likely useless, since they point to data from `argparse`.

Try this example which tries to simulate what we have here:

`pip install memory_profiler`

```python
from memory_profiler import profile


class A:
    def __init__(self):
        self.v = "a" * 1000000


class B:
    def __init__(self, obj):
        self._v = obj.v

    def clear_v(self):
        del self._v


@profile
def main():
    obj_a = A()
    obj_b = B(obj_a)
    obj_b.clear_v()
    del obj_b
    del obj_a


if __name__ == '__main__':
    main()
```

```
Line #    Mem usage    Increment  Occurrences   Line Contents
=============================================================
    17     35.2 MiB     35.2 MiB           1   @profile
    18                                         def main():
    19     36.3 MiB      1.0 MiB           1       obj_a = A()
    20     36.3 MiB      0.0 MiB           1       obj_b = B(obj_a)
    21     36.3 MiB      0.0 MiB           1       obj_b.clear_v()
    22     36.3 MiB      0.0 MiB           1       del obj_b
    23     35.4 MiB     -0.9 MiB           1       del obj_a
```

It is important to understand that Python only copies data if needed, and it uses references to this data.
In the case here, when you write `self._exec = argparse_args.exec` you just create a reference to the data initially stored by `argparse_args.exec`, so calling `del self._exec` don't actually free the memory used by the data, at least until you also `del argparse_args.exec`.

If we'd go for memory usage optimization it is rather elsewhere I'd look first, mainly in all the code processing data from MB or cover art for (possibly) millions of tracks... The code here is executed only once.

Note that I'm not against memory usage optimization, but to be clear, Python isn't the language for that... and if you really want to do that, use a memory profiler.

**Review by zas — CHANGES_REQUESTED** [review]

**Review by zas — CHANGES_REQUESTED** [review]

**Comment by zas** [comment]
See https://github.com/skelly37/picard/pull/8

**Comment by rdswift** [comment]
See https://github.com/skelly37/picard/pull/9

File: picard/tagger.py
Comment (zas) [inline picard/tagger.py:211]: Minor thing, but replacement  for `"%r" % obj` is `f"{repr(obj)}` (or `f"{obj!r}'`).
Not sure this change was intentional, in many cases the result is the same, but not always.

But anyway using Python3 formatted strings is better, so +1 from me.

Comment (skelly37) [reply picard/tagger.py:211]: Right, thanks

File: picard/tagger.py
Comment (skelly37) [reply picard/tagger.py:1288]: Really well explained and helpful, thanks

File: picard/tagger.py
Comment (zas) [reply picard/tagger.py:1408]: @skelly37 we may add `# nosec: B303` to this line, here `md5()` use is intended.

File: picard/tagger.py
Comment (skelly37) [reply picard/tagger.py:806]: So should I revert all the f-string refactors and then we can merge?

File: picard/util/pipe.py
Comment (zas) [inline picard/util/pipe.py:142]: @rdswift can `identifier` contain spaces the way it is build atm?

File: picard/util/pipe.py
Comment (zas) [inline picard/util/pipe.py:197]: For extra safety we could wrap filename with `sanitize_filename()` (which replaces path separators).

**Review by zas — APPROVED** [review]

File: picard/tagger.py
Comment (skelly37) [reply picard/tagger.py:806]: btw, a suggestion+question

Is there any reason why resolving all conversations is not required before merging? See: 
image

Comment (zas) [reply picard/tagger.py:806]: Keep them, those are minor changes, but next time try to keep unrelated changes outside ;)

File: picard/util/pipe.py
Comment (skelly37) [reply picard/util/pipe.py:197]: You mean this? https://pypi.org/project/sanitize-filename/

File: picard/util/pipe.py
Comment (rdswift) [reply picard/util/pipe.py:142]: No.  Good catch.  I had originally included the `.replace()` because I was first passing the whole config file path (which could contain spaces) but then decided to use the hash and forgot to remove the `.replace()`.

Comment (skelly37) [reply picard/util/pipe.py:142]: I don't see any way to add spaces to the identifier but I think it is a good security feature that may prevent future errors.

Comment (skelly37) [reply picard/util/pipe.py:142]: Or we can simply add a test case for pipe where we would assert that this variable contains no spaces, which gives better feedback for the developer

File: picard/tagger.py
Comment (skelly37) [reply picard/tagger.py:806]: > btw, a suggestion+question
> 
> Is there any reason why resolving all conversations is not required before merging? See: image

ping @zas to ensure that you have seen the question (as the time difference is small and you could've not refresh).

File: picard/util/pipe.py
Comment (skelly37) [reply picard/util/pipe.py:142]: I mean, the question is up to you @rdswift 

Could whitespaces in this string potentially break Picard?

If yes, I will add a proper test case. If not, I will remove the `.replace()` call :)

File: picard/util/pipe.py
Comment (rdswift) [reply picard/util/pipe.py:197]: Nope.  See Line 314 in `util/__init__.py`.

File: picard/util/pipe.py
Comment (rdswift) [reply picard/util/pipe.py:142]: I actually added it because I wasn't sure if a space would break the pipe name.  :stuck_out_tongue_winking_eye:   If that's not a concern, then there's likely no reason to keep it.

File: picard/util/pipe.py
Comment (skelly37) [reply picard/util/pipe.py:197]: > Nope. See Line 314 in `util/__init__.py`.

Thanks

---

So you suggest to call the sanitizer on the `os.path.join()` result, right?

Comment (zas) [reply picard/util/pipe.py:197]:
```suggestion
                                                sanitize_filename(f"{app_name}_v{app_version}_{self._identifier}_pipe_file")))
```

Comment (rdswift) [reply picard/util/pipe.py:197]: I *think* he's suggesting to wrap `f"{app_name}_v{app_version}_{self._identifier}_pipe_file"`.  If you whap the whole join it would remove the path separators from the name that you're appending.

File: picard/tagger.py
Comment (zas) [reply picard/tagger.py:806]: Well, no particular reason, I can enable it, and if it becomes annoying we can always disable it.
@phw^^

File: picard/util/pipe.py
Comment (zas) [reply picard/util/pipe.py]: > So you suggest to call the sanitizer on the `os.path.join()` result, right?

Nope, only on the filename, see suggestion above.

File: picard/util/pipe.py
Comment (zas) [inline picard/util/pipe.py:198]: You need to import it ;)

**Review by zas — CHANGES_REQUESTED** [review]

File: picard/util/pipe.py
Comment (skelly37) [reply picard/util/pipe.py:198]: Already done ;) https://github.com/metabrainz/picard/pull/2139/commits/6c510ed814b77fa9d6e7b4f6387bfef60a0d7ad5

**Comment by skelly37** [comment]
Sorry, I'm not on my main machine and always forget to set up the pre-commit hook on machines for journey...

**Review by zas — APPROVED** [review]
LGTM

**Review by rdswift — COMMENTED** [review]
This is perhaps a really minor point, but does the comment at Line 322 of `tagger.py` still apply?

**Comment by skelly37** [comment]
> This is perhaps a really minor point, but does the comment at Line 322 of `tagger.py` still apply? Otherwise, looks good to me.

You are right, it should be removed, thanks for noticing. And imho, it's not *a really minor point*. Outdated comments can cause a big mess and I strongly advise everyone not to ignore them.

**Review by rdswift — APPROVED** [review]

**Comment by skelly37** [comment]
These MacOS 3.7 tests (and only them) tend to break rarely. I think this should be researched some time.

**Comment by skelly37** [comment]
picard --help also needs to be cleaned from (starts a stand-alone instance) artifacts. I have just noticed it so please do not merge before such commit :)

**Review by phw — DISMISSED** [review]
This is a good improvement. The code looks cleaner that way again.

Given the changes I think this could be put into a single commit, we should use squash on merge here on Github.

**Comment by skelly37** [comment]
@rdswift please take a quick look if I have documented your changes fully correctly in the last commit :)

**Review by rdswift — APPROVED** [review]
The help changes look good to me.  Well done.

**Review by zas — APPROVED** [review]

**Comment by zas** [comment]
Grrr, I rebased instead of squashing commits. Well, that's done.

**Pull request merged**
